### ISSUE TITLE
GCC Optimisation Flags for Wine (-Os VS -O2)

### ISSUE THREAD

**Issue by Enverex** [Tk-Glitch/PKGBUILDS#61]
(couldn't find any discussion about this here so apologies if there is and I missed it)

Wine, by default, used to use O2 to build by default and it's also the Arch default. I've been using O3 with Wine for years (as the additional binary size overhead isn't an issue on modern processors) but it seems that the TKG version of Wine uses Os for some reason which seems like an odd choice given that it's normally only used for embedded systems or processors with very small internal cache.

TL;DR: What's the reason for using Os rather than O2 or O3?

**Comment by openglfreak** [comment]
-Os is probably faster because of cache.  
-O3 shouldn't be used for wine because the aggressive optimizations could break something.

**Comment by Tk-Glitch** [comment]
GCC's -O3 is almost always slower than -O2. I was using -O2 for wine-tkg before someone sent me benchmarks of -O2 vs -Os with Wine on a Sandy Bridge i3, showing -Os was usually faster, sometimes equal, and never slower. I have done a few gaming benchmarks myself after that on a Westmere Xeon with similar results. I have yet to do any measurement on a Coffee Lake i7 and I don't have access to a Zen CPU, but I'd be surprised if there was a big difference there.

Also, in the best case and on the machines and wine version tested (3.17-staging if my memory serves me well), the perf delta between -O2 and -Os was at best around 0.7% in favor of -Os. Nothing groundbreaking.

**Comment by Enverex** [comment]
Thanks for the explanation. I assumed there was good reason for changing from the defaults but couldn't find anything. I'll close this now.

**Comment by clapbr** [comment]
interesting info, did you make same tests for kernel and mesa? I have been compiling both with -O3 but this is not the first time I hear -O2 is faster.

**Comment by Tk-Glitch** [comment]
I've done extensive testing on Snapdragon based platforms (mostly custom Cortex-A8 and A9 based ones) regarding the kernel some years back, and `-O2` was faster than `-O3` on some SoCs, but not all.
That's probably just related to the quality of the platform implementation code-wise though. Considering at what pace new SoCs were produced at the time, and how badly optimized or even supported were these platforms, I'm pretty sure `-O3` was just hitting roadblocks over roadblocks in the slow cases.

On x86_64 however, since the code is much cleaner and well maintained, `-O3` should be pretty good and faster than `-O2` in various situations. `-O2` with `-march=native` might still be a better all-arounder though.. I'd need to update myself on the matter to be honest, so I won't draw any conclusion on that one.

I'm more or less convinced about mesa not liking -O3 though. I do remember some benchmarks about this where `-O3` was always slower than `-O2`.

As a side note, if by any chance someone finds a way to build wine with LTO I'm definitely interested :smiley:

**Comment by Enverex** [comment]
My only complaint with those as a basis for tests is that they have very small caches compared to decent desktop processors (e.g. i5/i7 machines) where Os makes more sense.

Phoronix did some tests relatively recently (not Wine based unfortunately though) - https://www.phoronix.com/scan.php?page=article&item=clear-gcc6-opts&num=1

**Comment by clapbr** [comment]
> As a side note, if by any chance someone finds a way to build wine with LTO I'm definitely interested
I have recently tried to build mesa with LTO without much luck. Could get rid of a few errors manually but it is too much.

**Comment by Tk-Glitch** [comment]
Interestingly I didn't try -Os on ARM yet :smile:

**Comment by ryanmusante** [comment]
What of using system-wide -Os? Is this an impediment or a help to my other packages like gzdoom-git/ mpv-git/ openvpn-git?
```
CFLAGS="-march=native -pipe -Os -fstack-protector-strong"
CXXFLAGS="${CFLAGS}"
LDFLAGS="-Wl,-Os,--sort-common,--as-needed,-z,relro,-z,now"

```

**Comment by Tk-Glitch** [comment]
This is not good for everything. I wouldn't use that globally. Basically what Enverex said.
